File: epsffit.c
/* epsffit.c
 * Fit EPSF file into constrained size
 *
 * (c) Reuben Thomas 2012-2016
 * (c) Angus J. C. Duggan 1991-1997
 * See file LICENSE for details.
 *
 * Added filename spec (from Larry Weissman) 5 Feb 93
 * Accepts double %%BoundingBox input, outputs proper BB, 4 Jun 93. (I don't
 * like this; developers should read the Big Red Book before writing code which
 * outputs PostScript.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include "progname.h"
#include "binary-io.h"
#include "minmax.h"

#include "psutil.h"
#include "psspec.h"

const char *syntax = "[-c] [-r] [-a] [-m] [-s] LLX LLY URX URY [INFILE [OUTFILE]]";

const char *argerr_message = "bad dimension";

int
main(int argc, char **argv)
{
   int bbfound = 0;              /* %%BoundingBox: found */
   int urx = 0, ury = 0, llx = 0, lly = 0;
   double furx, fury, fllx, flly;
   int showpage = 0, centre = 0, rotate = 0, aspect = 0, maximise = 0;
   char buf[BUFSIZ];
   int opt;

   set_program_name(argv[0]);

   while((opt = getopt(argc, argv, "csramv")) != EOF) {
     switch(opt) {
     case 'c': centre = 1; break;
     case 's': showpage = 1; break;
     case 'r': rotate = 1; break;
     case 'a': aspect = 1; break;
     case 'm': maximise = 1; break;
     case 'v':
     default:
       usage();
       break;
     }
   }

   if ((argc - optind) < 4)
      usage();

   fllx = singledimen(argv[optind++]);
   flly = singledimen(argv[optind++]);
   furx = singledimen(argv[optind++]);
   fury = singledimen(argv[optind++]);

   parse_input_and_output_files(argc, argv, optind, 0);

   while (fgets(buf, BUFSIZ, infile)) {
      if (buf[0] == '%' && (buf[1] == '%' || buf[1] == '!')) {
	 /* still in comment section */
	 if (!strncmp(buf, "%%BoundingBox:", 14)) {
	    double illx, illy, iurx, iury;	/* input bbox parameters */
	    if (sscanf(buf, "%%%%BoundingBox:%lf %lf %lf %lf\n",
		       &illx, &illy, &iurx, &iury) == 4) {
	       bbfound = 1;
	       llx = (int)illx;	/* accept doubles, but convert to int */
	       lly = (int)illy;
	       urx = (int)(iurx+0.5);
	       ury = (int)(iury+0.5);
	    }
	 } else if (!strncmp(buf, "%%EndComments", 13)) {
	    strcpy(buf, "\n"); /* don't repeat %%EndComments */
	    break;
	 } else writestring(buf);
      } else break;
   }

   if (!bbfound)
      die("no %%%%BoundingBox:");

   /* put BB, followed by scale&translate */
   double xoffset = fllx, yoffset = flly;
   double width = urx-llx, height = ury-lly;

   if (maximise)
      if ((width > height && fury-flly > furx - fllx) ||
          (width < height && fury-flly < furx - fllx))
         rotate = 1;

   int fwidth, fheight;
   if (rotate) {
      fwidth = fury - flly;
      fheight = furx - fllx;
   } else {
      fwidth = furx - fllx;
      fheight = fury - flly;
   }

   double xscale = fwidth/width;
   double yscale = fheight/height;

   if (!aspect)         /* preserve aspect ratio ? */
      xscale = yscale = MIN(xscale,yscale);
   width *= xscale;     /* actual width and height after scaling */
   height *= yscale;
   if (centre) {
      if (rotate) {
         xoffset += (fheight - height)/2;
         yoffset += (fwidth - width)/2;
      } else {
         xoffset += (fwidth - width)/2;
         yoffset += (fheight - height)/2;
      }
   }
   writestringf("%%%%BoundingBox: %d %d %d %d\n", (int)xoffset, (int)yoffset,
           (int)(xoffset + (rotate ? height : width)),
           (int)(yoffset + (rotate ? width : height)));
   if (rotate) {  /* compensate for original image shift */
      xoffset += height + lly * yscale;  /* displacement for rotation */
      yoffset -= llx * xscale;
   } else {
      xoffset -= llx * xscale;
      yoffset -= lly * yscale;
   }
   writestring("%%EndComments\n");
   if (showpage)
      writestring("save /showpage{}def /copypage{}def /erasepage{}def\n");
   else
      writestring("%%BeginProcSet: epsffit 1 0\n");
   writestring("gsave\n");
   writestringf("%.3f %.3f translate\n", xoffset, yoffset);
   if (rotate)
      writestring("90 rotate\n");
   writestringf("%.3f %.3f scale\n", xscale, yscale);
   if (!showpage)
      writestring("%%EndProcSet\n");
   do {
      writestring(buf);
   } while (fgets(buf, BUFSIZ, infile));
   writestring("grestore\n");
   if (showpage)
      writestring("restore showpage\n"); /* just in case */

   return 0;
}

File: psutils/epsffit.py
import importlib.metadata

VERSION = importlib.metadata.version('psutils')

version_banner=f'''\
%(prog)s {VERSION}
Copyright (c) Reuben Thomas 2023.
Released under the GPL version 3, or (at your option) any later version.
'''

import argparse
import re
import sys
import warnings
from typing import List

from psutils import (
    HelpFormatter, die, parsedimen, setup_input_and_output, simple_warning,
)

def get_parser() -> argparse.ArgumentParser:
    # Command-line arguments
    parser = argparse.ArgumentParser(
        description='Fit an Encapsulated PostScript file to a given bounding box.',
        formatter_class=HelpFormatter,
        usage='%(prog)s [OPTION...] LLX LLY URX URY [INFILE [OUTFILE]]',
        add_help=False,
    )
    warnings.showwarning = simple_warning(parser.prog)

    parser.add_argument('-c', '--center', '--centre', action='store_true',
                        help='center the image in the given bounding box')
    parser.add_argument('-r', '--rotate', action='store_true',
                        help='rotate the image by 90 degrees counter-clockwise')
    parser.add_argument('-a', '--aspect', action='store_true',
                        help='adjust the aspect ratio to fit the bounding box')
    parser.add_argument('-m', '--maximize', '--maximise', action='store_true',
                        help='rotate the image to fill more of the page if possible')
    parser.add_argument('-s', '--showpage', action='store_true',
                        help='append a /showpage to the file to force printing')
    parser.add_argument('--help', action='help',
                        help='show this help message and exit')
    parser.add_argument('-v', '--version', action='version',
                        version=version_banner)
    parser.add_argument('fllx', metavar='LLX', type=parsedimen,
                        help='x coordinate of lower left corner of the box')
    parser.add_argument('flly', metavar='LLX', type=parsedimen,
                        help='y coordinate of lower left corner of the box')
    parser.add_argument('furx', metavar='LLX', type=parsedimen,
                        help='x coordinate of upper right corner of the box')
    parser.add_argument('fury', metavar='LLX', type=parsedimen,
                        help='y coordinate of upper right corner of the box')
    parser.add_argument('infile', metavar='INFILE', nargs='?',
                        help="`-' or no INFILE argument means standard input")
    parser.add_argument('outfile', metavar='OUTFILE', nargs='?',
                        help="`-' or no OUTFILE argument means standard output")
    return parser

def epsffit(argv: List[str]=sys.argv[1:]) -> None: # pylint: disable=dangerous-default-value
    args = get_parser().parse_intermixed_args(argv)

    urx, ury, llx, lly = 0, 0, 0, 0

    infile, file_type, outfile = setup_input_and_output(args.infile, args.outfile)
    if file_type not in ('.ps', '.eps'):
        die(f"incompatible file type `{args.infile}'")

    def output(s: str) -> None:
        outfile.write((s + '\n').encode('utf-8'))

    bbfound = False # %%BoundingBox: found
    for line in infile:
        if re.match(b'%[%!]', line): # still in comment section
            if line.startswith(b'%%BoundingBox:'):
                m = re.match(b'%%BoundingBox: +([\\d.]+) +([\\d.]+) +([\\d.]+) +([\\d.]+)$', line)
                if m:
                    bbfound = True
                    llx = int(m[1]) # accept doubles, but convert to int
                    lly = int(m[2])
                    urx = int(float(m[3]) + 0.5)
                    ury = int(float(m[4]) + 0.5)
            elif line.startswith(b'%%EndComments'): # don't repeat %%EndComments
                break
            else:
                outfile.write(line)
        else:
            break

    if bbfound is False:
        die('no %%BoundingBox:', 2)

    # Write bounding box, followed by scale & translate
    xoffset, yoffset = args.fllx, args.flly
    width, height = urx - llx, ury - lly

    # FIXME: Consider more carefully how --rotate and --maximize should interact
    rotate = args.rotate
    if args.maximize and \
        (((width > height) and (args.fury - args.flly > args.furx - args.fllx)) or \
        ((width < height) and (args.fury - args.flly < args.furx - args.fllx))):
        rotate = True

    fwidth, fheight = args.furx - args.fllx, args.fury - args.flly
    if rotate:
        fwidth, fheight = fheight, fwidth

    xscale, yscale = fwidth / width, fheight / height

    if not args.aspect: # preserve aspect ratio?
        xscale = yscale = min(xscale, yscale)
    width *= xscale # actual width and height after scaling
    height *= yscale
    if args.center:
        if rotate:
            xoffset += (fheight - height) / 2
            yoffset += (fwidth - width) / 2
        else:
            xoffset += (fwidth - width) / 2
            yoffset += (fheight - height) / 2
    output(f'%%BoundingBox: {int(xoffset)} {int(yoffset)} {int(xoffset + (height if rotate else width))} {int(yoffset + (width if rotate else height))}')
    if rotate: # compensate for original image shift
        xoffset += height + lly * yscale # displacement for rotation
        yoffset -= llx * xscale
    else:
        xoffset -= llx * xscale
        yoffset -= lly * yscale
    output('%%EndComments')
    if args.showpage:
        output('save /showpage{}def /copypage{}def /erasepage{}def')
    else:
        output('%%BeginProcSet: epsffit 1 0')
    output(f'gsave {xoffset:.3f} {yoffset:.3f} translate')
    if rotate:
        output('90 rotate')
    output(f'{xscale:.3f} {yscale:.3f} scale')
    if not args.showpage:
        output('%%EndProcSet')
    outfile.write(infile.read())
    output('grestore')
    if args.showpage:
        output('restore showpage') # just in case


if __name__ == '__main__':
    epsffit()
